### PR TITLE
Filter methods completion with Qute rules

### LINKED ISSUE #463: Filter methods completion with Qute rules
Method completion returns too many completion items which are not valid (ex : wait() method from object is not allowed). 

See rules at https://github.com/quarkusio/quarkus/blob/ce19ff75e9f732ff731bb30c2141b44b42c66050/independent-projects/qute/core/src/main/java/io/quarkus/qute/ReflectionValueResolver.java#L176

It should be nice to:

 * fix completion with those rule
 * report diagnostic to explain why wait is forbidden.

### PR DESCRIPTION
**Pull request by angelozerr**
Filter methods completion with Qute rules

Fixes #463

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
Here a demo:

InvalidMethodDemo

I need to write tests.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java [hunk 1]
```diff
@@ -397,6 +407,33 @@ private static boolean isValidMethod(IMethod method, boolean isInterface) {
...
+	 * 
+	 * @return
+	 * 
+	 * @see https://github.com/quarkusio/quarkus/blob/ce19ff75e9f732ff731bb30c2141b44b42c66050/independent-projects/qute/core/src/main/java/io/quarkus/qute/ReflectionValueResolver.java#L176
+	 */
+	private static InvalidMethodReason getValidMethodForQute(IMethod method, String typeName) {
+		if (method.getNumberOfParameters() > 0) {
+			return InvalidMethodReason.HasParameters;
+		}
+		if ("java.lang.Object".equals(typeName)) {
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java:424]: Move Object methods check first

Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java [hunk 2]
```diff
@@ -27,6 +27,10 @@
 	UndefinedVariable("`{0}` cannot be resolved to a variable."), //
 	UnkwownType("`{0}` cannot be resolved to a type."), //
 	UnkwownMethod("`{0}` cannot be resolved or is not a method of `{1}` Java type."), //
+	InvalidMethodVoid("Invalid `{0}` method of `{1}` Java type : void return is not allowed."), //
+	InvalidMethodHasParameters("Invalid `{0}` method of `{1}` Java type : method with parameters is not allowed."), //
+	InvalidMethodFromObject("Invalid `{0}` method of `{1}` Java type : method from java.lang.Object is not allowed."), //
```
Comment (fbricon) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java:32]: Can we drop the "Java type" parts? Feels verbose

Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java]: fixed

**Comment by fbricon** [comment]
mmm so, given 
```java
public class Item {
    public BigDecimal price;
    public String name;
    public BigDecimal getDiscountedPrice(String percent) {
        return price.subtract(price.multiply(new BigDecimal(percent)));
    }
}
```

I get this error:
Screenshot 2021-12-10 at 12 15 04

But Quarkus runs fine with it, computes and displays the discounted price.

@mkouba isn't Quarkus supposed to prevent using parameters with https://github.com/quarkusio/quarkus/blob/ce19ff75e9f732ff731bb30c2141b44b42c66050/independent-projects/qute/core/src/main/java/io/quarkus/qute/ReflectionValueResolver.java#L178?

**Comment by angelozerr** [comment]
It means that we must remove the check with has parameters?

**Comment by mkouba** [comment]
> It means that we must remove the check with has parameters?

More or less ;-). The `ReflectionValueResolver` is a fallback that works fine in JVM but for native image you'd have to register the classes for reflection. On the other hand, if a value resolver is generated for a class (e.g. all types used in all type-safe expressions) then the `ReflectionValueResolver` is not needed at all (neither in JVM nor for native image).

**Comment by mkouba** [comment]
Note that in Quarkus, you can also define [template extension methods](https://quarkus.io/guides/qute-reference#template_extension_methods) that can be used to extend the set of accessible properties and methods. So it doesn't have to a class property/method. These are not available in qute standalone though.

**Comment by angelozerr** [comment]
> Note that in Quarkus, you can also define template extension methods that can be used to extend the set of accessible properties and methods. So it doesn't have to a class property/method. 

Thanks for the info, but It's already supported :) This issue is about method rules. I think I will remove hasParameters check for now. @fbricon are you OK?

**Comment by fbricon** [comment]
yeah remove it. 

And generally speaking, it's easier to be more lenient than the real Quarkus parser, we'll hear less complaints in that way

**Comment by fbricon** [comment]
We still get a *ton of noise from collection methods for instance:

Screenshot 2021-12-13 at 12 46 14

Not sure how we're supposed to filter them out.

**Comment by angelozerr** [comment]
> yeah remove it.

I did that.

> We still get a *ton of noise from collection methods for instance:

I don't know how to filter those methods, but a thing which is sure is that I don't support generic with those ethod(for completion, validation). But I think it's ananother issue.

**Comment by angelozerr** [comment]
@mkouba it seems that we cannot use static method List.of.  Is it true?

**Comment by mkouba** [comment]
> @mkouba it seems that we cannot use static method List.of. Is it true?

Yes, static members can only be accessed through a namespace or custom template extension methods. See also [Accessing Static Fields and Methods](https://quarkus.io/guides/qute-reference#accessing-static-fields-and-methods).

**Comment by angelozerr** [comment]
> Yes, static members can only be accessed through a namespace or custom template extension methods. See also Accessing Static Fields and Methods.

Thanks for your information!

**Comment by angelozerr** [comment]
@fbricon now I set static method as an error. 

When static method is annotated with TemplateExtension it should work. It misses the case with static method syantx with package name.

**Comment by mkouba** [comment]
> Is that a bug?

I can't see any [template extension method](https://quarkus.io/guides/qute-reference#template_extension_methods) in your snippet. Anyway, I think that we don't handle static methods in the generated value resolver but it seems that we allow invocation of static methods via the fallback reflection resolver. In general, you can invoke a static method upon an instance in java too ...although you get a warning from the compiler and it's a bad practice. Also I think that the type-safe validation check would fail in this case. So perhaps we should remove the support for this kind of incovation from the `ReflectionValueResolver`. WDYT?

**Comment by angelozerr** [comment]
> WDYT?

To be consistent and avoid some confuse, I think it should be safe to removethis kind of invocation. In other words,forbid all use of static methods.

**Comment by mkouba** [comment]
> > WDYT?
> 
> To be consistent and avoid some confuse, I think it should be safe to removethis kind of invocation. In other words,forbid all use of static methods (exepct with TemplateExtension or package name).

Ok, I'll send a PR.

**Comment by fbricon** [comment]
Thanks @angelozerr

**Comment by mkouba** [comment]
> > Ok, I'll send a PR.
> 
> Thank so much!

https://github.com/quarkusio/quarkus/pull/22233

**Pull request merged**